kgl-math/src/commonMain/kotlin/com/kgl/math/Vectors.kt

```diff
@@ -0,0 +1,289 @@
+package com.kgl.math
+
+/**
+ * A generic two-component vector. Methods in this interface support only read-only access to the vector;
+ * read/write access is supported through the [MutableVector2] interface.
+ * @param T the type of component contained in the vector. The vector is invariant on its component type.
+ */
+interface Vector2<T> {
```

Comment:
For the sake of preventing auto-boxing, I think we should reify these by hand. Just `Vector2f` for now.

Reply:
To clarify: by reify by hand, what exactly are you envisioning?

Reply:
Instead of having `interface Vector2<T> { ... }`, allowing implementations like `Vector2<Float>`, `Vector2<Double>`, `Vector2<Int>`, `Vector2<CustomUserNumber>`, etc.

We should just have `interface Vector2f { ... }`, `interface Vector2d { ... }`, `interface Vector2i { ... }`, etc. Hard coding the component type, instead of using generics.

Reply:
That’s what I thought. The reason I’m using generics is mainly for code reuse. My local changes have changed from interface to sealed class. Will that have the same effect, do you think?

Reply:
The code reuse point is a strong one for generics. Might be worth considering code gen (or whatever `kotlin-stdlib` does for `map`, `filter`, etc) for reification, later on.

Reply:
In what case would auto-boxing occur? If we use a sealed class with only the primitive types and Boolean defined as subclasses, would that not also prevent boxing in general? (other than the obvious array storage)

Reply:
Auto-boxing would occur anywhere `T` is passed as a parameter or returned. Like when calling `x`, `y` and especially in `equals` (since it gets all the components).

Reply:
I see. So this is not the kind of auto-boxing that we see with nullable primitives... Codegen definitely seems the best approach, although I wanna do some testing with sealed classes to check if that still causes heap allocations.

Reply:
converted to codegen. Currently configured to generate Float and Double variants.